### PR TITLE
UF-7471: Handle attachments from e-service and provide them in the API

### PR DESCRIPTION
**Pull request by cheezi0747**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -110,8 +113,11 @@
 					<excludes>
 						<exclude>**/generated/**</exclude>
 						<exclude>**/se/sundsvall/webmessagecollector/Application*</exclude>
-						<exclude>**/se/sundsvall/webmessagecollector/integration/db/model/**/*_.*</exclude>
-						<exclude>**/se/sundsvall/webmessagecollector/integration/opene/OpenEClient.*</exclude>
+						<exclude>**/se/sundsvall/webmessagecollector/integration/db/model/**/*_.*
+						</exclude>
+						<exclude>
+							**/se/sundsvall/webmessagecollector/integration/opene/OpenEClient.*
```
Comment (foobar-1337) [inline pom.xml:119]: Why are these files manually coded? Isn't it possible to  generate them?

(If we still want to "own" the source code for the client model, we should at least have tests for them, i.m.h.o)

File: src/main/java/se/sundsvall/webmessagecollector/api/MessageResource.java [hunk 2]
```diff
@@ -38,19 +41,38 @@ public MessageResource(final MessageService service) {
...
+	@ApiResponse(responseCode = "200", description = "Successful Operation", useReturnTypeSchema = true)
+	@Operation(summary = "Get a messageAttachment", description = "Returns a messageAttachment as a BASE64 string for the specified attachmentId")
+	public ResponseEntity<String> getAttachment(
+		@Parameter(name = "attachmentId", description = "MessageId to fetch attachment for", example = "123", required = true) @PathVariable("attachmentId") final int attachmentId) {
+		return ResponseEntity.ok(service.getAttachment(attachmentId));
 	}
+
+	@DeleteMapping("attachments/{attachmentId}")
+	@ApiResponse(responseCode = "204", description = "No Content", useReturnTypeSchema = true)
+	@Operation(summary = "Delete a messageAttachment", description = "Deletes a messageAttachment with the specified id")
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/webmessagecollector/api/MessageResource.java:70]: "Delete a messageAttachment" 
->
"Delete a message attachment" 

?

File: src/main/java/se/sundsvall/webmessagecollector/api/model/MessageAttachment.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+@Builder(setterPrefix = "with")
+public class MessageAttachment {
+
+	@Schema(description = "The Id for the attachment", example = "1")
+	private final Integer attachmentId;
+
+	@Schema(description = "The name of the file", example = "file.txt")
+	private final String fileName;
+
+	@Schema(description = "The file as a base64 encoded string", example = "dGhpcyBpcyBhIGJhc2U2NCBlbmNvZGVkIHN0cmluZw==")
```
Comment (ronnberg) [inline src/main/java/se/sundsvall/webmessagecollector/api/model/MessageAttachment.java:17]: Do we need the file represented as base64 string? Why not use streams?

File: pom.xml [hunk 4]
```diff
@@ -110,8 +113,11 @@
 					<excludes>
 						<exclude>**/generated/**</exclude>
 						<exclude>**/se/sundsvall/webmessagecollector/Application*</exclude>
-						<exclude>**/se/sundsvall/webmessagecollector/integration/db/model/**/*_.*</exclude>
```
Comment (foobar-1337) [inline pom.xml:113]: It's possible to create tests for the DB-model files

File: src/main/java/se/sundsvall/webmessagecollector/integration/db/MessageAttachmentRepository.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package se.sundsvall.webmessagecollector.integration.db;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import se.sundsvall.webmessagecollector.integration.db.model.MessageAttachmentEntity;
+
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/webmessagecollector/integration/db/MessageAttachmentRepository.java:6]: Add circuit-breaker?

File: src/test/java/se/sundsvall/webmessagecollector/integration/db/model/MessageAttachmentEntityTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		final var result = MessageAttachmentEntity.builder()
+			.withFile(file)
+			.withAttachmentId(attachmentId)
+			.withFileName(fileName)
+			.build();
+		// Assert
+		assertThat(result).isNotNull().hasNoNullFieldsOrProperties();
+		assertThat(result.getAttachmentId()).isEqualTo(attachmentId);
+		assertThat(result.getFileName()).isEqualTo(fileName);
+		assertThat(result.getFile()).isEqualTo(file);
```
Comment (ronnberg) [inline src/test/java/se/sundsvall/webmessagecollector/integration/db/model/MessageAttachmentEntityTest.java:48]: isSameAs insted of isEqualsTo?

File: src/test/java/se/sundsvall/webmessagecollector/integration/db/model/MessageEntityTest.java [hunk 7]
```diff
@@ -77,5 +80,7 @@ void testBuildPattern() {
 		assertThat(bean.getSent()).isEqualTo(sent);
 		assertThat(bean.getUserId()).isEqualTo(userId);
 		assertThat(bean.getUsername()).isEqualTo(username);
+		assertThat(bean.getAttachments()).isEqualTo(attachments);
```
Comment (ronnberg) [inline src/test/java/se/sundsvall/webmessagecollector/integration/db/model/MessageEntityTest.java:83]: isSameAs?

File: src/test/java/se/sundsvall/webmessagecollector/service/MessageServiceTest.java [hunk 8]
```diff
@@ -45,20 +58,21 @@ void getMessages() {
...
 		// Assert
 		assertThat(result).isNotNull().hasSize(2);
 		assertThat(result.getFirst()).hasNoNullFieldsOrProperties();
 		assertThat(result.getFirst()).usingRecursiveComparison()
-			.isEqualTo(MessageMapper.toDTO(entity));
+			.isEqualTo(MessageMapper.toMessageDTO(entity));
 		// Verify
-		verify(repository).findAllByFamilyId(anyString());
-		verifyNoMoreInteractions(repository);
+		verify(messageRepository).findAllByFamilyId(anyString());
```
Comment (ronnberg) [inline src/test/java/se/sundsvall/webmessagecollector/service/MessageServiceTest.java:74]: anyString() -> "someFamilyId"

File: src/test/java/se/sundsvall/webmessagecollector/service/MessageServiceTest.java [hunk 9]
```diff
@@ -67,16 +81,85 @@ void getMessages_EmptyList() {
 		final var result = service.getMessages("someFamilyId");
 		// Assert & Verify
 		assertThat(result).isNotNull().isEmpty();
-		verify(repository).findAllByFamilyId(anyString());
-		verifyNoMoreInteractions(repository);
+		verify(messageRepository).findAllByFamilyId(anyString());
```
Comment (ronnberg) [inline src/test/java/se/sundsvall/webmessagecollector/service/MessageServiceTest.java:84]: anyString() -> "someFamilyId"

File: src/test/java/se/sundsvall/webmessagecollector/service/MessageServiceTest.java [hunk 10]
```diff
@@ -67,16 +81,85 @@ void getMessages_EmptyList() {
...
 	}
 
 	@Test
 	void deleteMessages() {
 		//Act
 		service.deleteMessages(List.of(1));
 		//Verify
-		verify(repository).deleteAllById(anyList());
-		verifyNoMoreInteractions(repository);
+		verify(messageRepository).deleteAllById(anyList());
```
Comment (ronnberg) [inline src/test/java/se/sundsvall/webmessagecollector/service/MessageServiceTest.java:93]: Verification should not be on anyList. Store a referense of the list and use same()

File: src/test/java/se/sundsvall/webmessagecollector/service/scheduler/MessageCacheServiceTest.java [hunk 11]
```diff
@@ -97,6 +104,10 @@ void fetchMessagesWithResponse() {
 			assertThat(entity.getSent()).isEqualTo(LocalDateTime.of(2022, 5, 25, 11, 20));
 			assertThat(entity.getUserId()).isNull();
 			assertThat(entity.getUsername()).isNull();
+			assertThat(entity.getAttachments()).hasSize(1);
+			assertThat(entity.getAttachments().getFirst().getAttachmentId()).isEqualTo(123);
+			assertThat(entity.getAttachments().getFirst().getFileName()).isEqualTo("someFile.pdf");
+			assertThat(entity.getAttachments().getFirst().getFile()).isNotNull();
```
Comment (ronnberg) [inline src/test/java/se/sundsvall/webmessagecollector/service/scheduler/MessageCacheServiceTest.java:110]: Change isNotNull to assert correct content?

File: src/main/java/se/sundsvall/webmessagecollector/service/scheduler/MessageCacheService.java [hunk 12]
```diff
@@ -38,15 +53,33 @@ public void fetchMessages(String familyId) {
 
 		final var bytes = openEClient.getMessages(familyId, fromTimestamp, "");
 		final var messages = toMessageEntities(bytes, familyId);
-		messageRepository.saveAll(messages);
 
+		messages.forEach(messageEntity -> {
```
Comment (ronnberg) [inline src/main/java/se/sundsvall/webmessagecollector/service/scheduler/MessageCacheService.java:57]: Is there risk that this transaction becomes too big? looping to fetch all attachments and storing every one in the same transaction seems like a risk for OOM.

**Review by foobar-1337 — APPROVED** [review]

File: src/main/java/se/sundsvall/webmessagecollector/service/scheduler/MessageCacheService.java [hunk 13]
```diff
@@ -41,12 +59,27 @@ public void fetchMessages(String familyId) {
...
 	}
 
-	private ExecutionInformationEntity initiateExecutionInfo(String familyId) {
+	@Transactional
+	public void fetchAttachment(final MessageAttachmentEntity attachmentEntity) {
+		try {
+			final var attachmentStream = openEClient.getAttachment(attachmentEntity.getAttachmentId());
+			if (attachmentStream != null) {
+				attachmentEntity.setFile(new SerialBlob(attachmentStream));
+				messageAttachmentRepository.save(attachmentEntity);
```
Comment (ronnberg) [inline src/main/java/se/sundsvall/webmessagecollector/service/scheduler/MessageCacheService.java:71]: I thing, if this is called from another transaction scope, that the session is not flushed when the method ends.. If I am not misstaking you should use saveAndFlush.

Comment (cheezi0747) [reply src/main/java/se/sundsvall/webmessagecollector/service/scheduler/MessageCacheService.java]: You are correct in this.

**Review by ronnberg — APPROVED** [review]

**Pull request merged**